### ISSUE TITLE
libai lr_scheduler设计文档

### ISSUE THREAD

**Issue by rentainhe** [Oneflow-Inc/libai#32]
### OneFlow中的Scheduler
oneflow中的scheduler在`eager`和`graph`下的计算逻辑是独立的，eager下的计算逻辑是写在python层里的，graph下的计算逻辑是在C++层面实现，oneflow为了保证在eager模式和graph模式下，都可以执行相同的scheduler，得到同样的结果，会在Scheduler中定义一个`_generate_conf_for_graph`方法，这个方法的作用仅仅是为了`初始化graph运行时的scheduler`，所以`初始化相同，C++计算逻辑和python计算逻辑相同`的情况下，可以保证我graph和eager模式下用的scheduler，得到的结果都是一样的

### Libai中的Scheduler设计讨论
libai最早的设计我是想仿照`detectron2`中的scheduler设计，detectron2中将scheduler的设计解耦成了以下几个部分：
- `ParamScheduler`：主要复杂根据程序执行的不同阶段，返回一个`scalar`
- `LRMultiplier`：继承自`flow.optim.lr_scheduler._LRScheduler`，主要需要传入的是optimizer参数以及`ParamScheduler`，然后在`get_lr`函数中，调用`ParamScheduler`，得到当前阶段的scalar，再把scalar乘上学习率

举一个简单的例子:
- `ConstantParamScheduler`：表示我训练的任何阶段学习率都是一样的，所以我只需要初始化这个ParamScheduler，然后重写`__call__`方法，在任意阶段返回相同的scalar即可
```python
class ConstantParamScheduler(ParamScheduler):
    """
    Returns a constant value for a param
    """

    def __init__(self, value: float) -> None:
        self._value = value
    
    def __call__(self, where: float) -> float:
        return self._value
```
- `LRMultiplier`：会在`get_lr`函数中传入当前运行的step，以及总的step数，用当前的step除以总的step表示训练的进度，再调用传入的`ParamScheduler`来得到当前阶段的scalar，再乘上学习率
```python
class LRMultiplier(flow.optim.lr_scheduler._LRScheduler):
    """
    A LRScheduler which uses :class:`ParamScheduler` to multiply the
    learning rate of each param in the optimizer.
    Every step, the learning rate of each parameter becomes its initial value
    multiplied by the output of the given :class:`ParamScheduler`.

    The absolute learning rate value of each parameter can be different.
    This scheduler can be used as long as the relative scale among them do
    not change during training.
    
    Examples:
    ::
        LRMultiplier(
            opt,
            WarmupParamScheduler(
                MultiStepParamScheduler(
                    [1, 0.1, 0.01],
                    milestones=[60000, 80000],
                    num_updates=90000,
                ), 0.001, 100 / 90000
            ),
            max_iter=90000
        )
    """

    # NOTES: in the most general case, every LR can use its own scheduler.
    # Supporting this requires interaction with the optimizer when its parameter
    # group is initialized. For example, classyvision implements its own optimizer
    # that allows different schedulers for every parameter group.
    # To avoid this complexity, we use this class to support the most common cases
    # where the relative scale among all LRs stay unchanged during training.  In this
    # case we only need a total of one scheduler that defines the relative LR multiplier.

    def __init__(
        self,
        optimizer: flow.optim.Optimizer,
        multiplier: ParamScheduler,
        max_iter: int,
        last_iter: int = -1,
    ):
        """
        Args:
            optimizer, last_iter: See ``flow.optim.lr_scheduler._LRScheduler``.
                ``last_iter`` is the same as ``last_epoch``.
            multiplier: a libai ParamScheduler that defines the multiplier on
                every LR of the optimizer
            max_iter: the total number of training iterations
        """
        if not isinstance(multiplier, ParamScheduler):
            raise ValueError(
                "_LRMultiplier(multiplier=) must be an instance of libai "
                f"ParamScheduler. Got {multiplier} instead."
            )
        self._multiplier = multiplier
        self._max_iter = max_iter
        super().__init__(optimizer, last_step=last_iter)
    
    def state_dict(self):
        # libai schedulers are stateless. Only keep oneflow scheduler states
        return {"base_lrs": self.base_lrs, "last_step": self.last_step}
    
    def get_lr(self) -> List[float]:
        multiplier = self._multiplier(self.last_step / self._max_iter)
        return [base_lr * multiplier for base_lr in self.base_lrs]
```

这种方案有个比较麻烦的点在于，我如果实现的scheduler需要同时兼容eager和graph的执行模式，我就必须要在`LRMultiplier`中定义一个`_generate_conf_for_graph`方法，并且在底层的`ParamScheduler`中也做同样的定义

举个例子
```python
class ConstantParamScheduler(ParamScheduler):
    """
    Returns a constant value for a param
    """

    def __init__(self, value: float) -> None:
        self._value = value
    
    def __call__(self, cur_iter: int, max_iter: int) -> float:
        return self._value
    
    def _generator_conf_for_graph(self, opt_confs):
        ...
```
```python
class LRMultiplier(flow.optim.lr_scheduler._LRScheduler):
    """
    A LRScheduler which uses :class:`ParamScheduler` to multiply the
    learning rate of each param in the optimizer.
    Every step, the learning rate of each parameter becomes its initial value
    multiplied by the output of the given :class:`ParamScheduler`.

    The absolute learning rate value of each parameter can be different.
    This scheduler can be used as long as the relative scale among them do
    not change during training.
    
    Examples:
    ::
        LRMultiplier(
            opt,
            WarmupParamScheduler(
                MultiStepParamScheduler(
                    [1, 0.1, 0.01],
                    milestones=[60000, 80000],
                    num_updates=90000,
                ), 0.001, 100 / 90000
            ),
            max_iter=90000
        )
    """

    # NOTES: in the most general case, every LR can use its own scheduler.
    # Supporting this requires interaction with the optimizer when its parameter
    # group is initialized. For example, classyvision implements its own optimizer
    # that allows different schedulers for every parameter group.
    # To avoid this complexity, we use this class to support the most common cases
    # where the relative scale among all LRs stay unchanged during training.  In this
    # case we only need a total of one scheduler that defines the relative LR multiplier.

    def __init__(
        self,
        optimizer: flow.optim.Optimizer,
        multiplier: ParamScheduler,
        max_iter: int,
        last_iter: int = -1,
    ):
        """
        Args:
            optimizer, last_iter: See ``flow.optim.lr_scheduler._LRScheduler``.
                ``last_iter`` is the same as ``last_epoch``.
            multiplier: a libai ParamScheduler that defines the multiplier on
                every LR of the optimizer
            max_iter: the total number of training iterations
        """
        if not isinstance(multiplier, ParamScheduler):
            raise ValueError(
                "_LRMultiplier(multiplier=) must be an instance of libai "
                f"ParamScheduler. Got {multiplier} instead."
            )
        self._multiplier = multiplier
        self._max_iter = max_iter
        super().__init__(optimizer, last_step=last_iter)
    
    def state_dict(self):
        # libai schedulers are stateless. Only keep oneflow scheduler states
        return {"base_lrs": self.base_lrs, "last_step": self.last_step}
    
    def get_lr(self) -> List[float]:
        multiplier = self._multiplier(self.last_step / self._max_iter)
        return [base_lr * multiplier for base_lr in self.base_lrs]
    
    # 在这里需要实现一个 _generate_conf_for_graph 方法
    def _generate_conf_for_graph(self, opt_confs):
        return self._multiplier._generate_conf_for_graph(opt_confs)
```
- 这么做会导致代码层层封装，不太方便使用

### Solution
个人感觉d2这种方案不太适合libai的场景，所以这里个人感觉还是采用oneflow内置的`Scheduler`，但是在`libai`里`重写一份`，为没有添加`_generate_conf_for_graph`的scheduler添加一个这个方法，但是`raise NotImplementedError("xxx scheduler only support eager mode now")`，添加一个报错信息方便用户使用
- 举个例子：在libai中重新copy一份LambdaLR
```python
import oneflow as flow

class LambdaLR(flow.optim.lr_scheduler._LRScheduler):
    def __init__(self, optimizer, lr_lambda, last_step=-1, verbose=False):
        if not isinstance(lr_lambda, (list, tuple)):
            self.lr_lambdas = [lr_lambda] * len(optimizer.param_groups)
        else:
            assert len(lr_lambda) == len(
                optimizer.param_groups
            ), f"Expected {len(optimizer.param_groups)} lr_lambdas, but got {len(lr_lambda)}"
            self.lr_lambdas = list(lr_lambda)
        super().__init__(optimizer, last_step, verbose)

    def state_dict(self):
        """Returns the state of the scheduler as a :class:`dict`.

        It contains an entry for every variable in self.__dict__ which
        is not the optimizer.
        The learning rate lambda functions will only be saved if they are callable objects
        and not if they are functions or lambdas.
        """
        state_dict = {
            key: value
            for (key, value) in self.__dict__.items()
            if key not in ("optimizer", "lr_lambdas")
        }
        state_dict["lr_lambdas"] = [None] * len(self.lr_lambdas)
        for (idx, fn) in enumerate(self.lr_lambdas):
            if not isinstance(fn, types.FunctionType):
                state_dict["lr_lambdas"][idx] = fn.__dict__.copy()
        return state_dict

    def load_state_dict(self, state_dict):
        """Loads the schedulers state.

        Arguments:
            state_dict (dict): scheduler state. Should be an object returned
                from a call to :meth:`state_dict`.
        """
        lr_lambdas = state_dict.pop("lr_lambdas")
        self.__dict__.update(state_dict)
        state_dict["lr_lambdas"] = lr_lambdas
        for (idx, fn) in enumerate(lr_lambdas):
            if fn is not None:
                self.lr_lambdas[idx].__dict__.update(fn)

    def get_lr(self):
        return [
            base_lr * lmbda(self.last_step)
            for (lmbda, base_lr) in zip(self.lr_lambdas, self.base_lrs)
        ]

    def _generate_conf_for_graph(self, opt_confs):
        # 添加报错信息，或者做一些别的操作
        raise NotImplementedError("LambdaLR scheduler only support eager mode now")
```
- 虽然从代码层面来说是和oneflow内置的几乎一样，但是我们可以拓展一些报错信息，以及统一一下Scheduler的命名规范，以及用户在使用的时候就可以直接`from libai.optim import CosineLRScheduler`，比较清晰

**Comment by CPFLAME** [comment]
graph确实存在这个问题,  采用自定义的lr, 必须要定义`_generate_conf_for_graph`, 而这个又需要调用C++的接口. 会导致用户完全定义不了自己的LR.
但是现在libai更多的是支持graph模式,  这就导致用户用libai 只能用oneflow内置的lr_scheduler

**Comment by L1aoXingyu** [comment]
我们可以通过一个 try catch 机制去捕获这个 error，不用重新 copy 一份一样的代码

**Comment by rentainhe** [comment]
> graph确实存在这个问题, 采用自定义的lr, 必须要定义`_generate_conf_for_graph`, 而这个又需要调用C++的接口. 会导致用户完全定义不了自己的LR. 但是现在libai更多的是支持graph模式, 这就导致用户用libai 只能用oneflow内置的lr_scheduler

可以的，写一个build来解决这个问题好了

**Comment by rentainhe** [comment]
> 我们可以通过一个 try catch 机制去捕获这个 error，不用重新 copy 一份一样的代码

OKKK

**Comment by dangkai4u** [comment]
先统计一下预计支持哪几种scheduler、目前可以支持哪几种（graph+eager）、暂不支持哪几种，列个清单。如果常用的都支持，个别可以不管。如果某些必要的scheduler目前不支持，那就在oneflow里添加一下吧。

**Comment by rentainhe** [comment]
> 先统计一下预计支持哪几种scheduler、目前可以支持哪几种（graph+eager）、暂不支持哪几种，列个清单。如果常用的都支持，个别可以不管。如果某些必要的scheduler目前不支持，那就在oneflow里添加一下吧。

暂时计划是先实现两个，`WarmupMultiStepLR`和`WarmupCosineLR`，满足目前训练的需求，后续再开PR添加其他的，可以统计一下，因为有些LR Scheduler可能几乎都用不上，不常见的让用户自己添加即可

**Comment by dangkai4u** [comment]
WarmupMultiStepLR是什么更新规则，感觉这些不够啊。transformer模型一般会用Inverse square root scheduler训练，用polynomial decay scheduler微调，这两个都支持warmup。咱们的模型主要是transformer，所以这两个应该也支持一下。

**Comment by rentainhe** [comment]
> WarmupMultiStepLR是什么更新规则，感觉这些不够啊。transformer模型一般会用Inverse square root scheduler训练，用polynomial decay scheduler微调，这两个都支持warmup。咱们的模型主要是transformer，所以这两个应该也支持一下。

可以，这些都很好支持，inverse square这个恐怕是要oneflow内部支持
